scrape/manager.go

```diff
@@ -50,28 +51,84 @@ type Manager struct {
 	append    Appendable
 	graceShut chan struct{}
 
-	mtxTargets     sync.Mutex // Guards the fields below.
-	targetsActive  []*Target
-	targetsDropped []*Target
-	targetsAll     map[string][]*Target
-
 	mtxScrape     sync.Mutex // Guards the fields below.
 	scrapeConfigs map[string]*config.ScrapeConfig
 	scrapePools   map[string]*scrapePool
+	targetSets    map[string][]*targetgroup.Group
+
+	triggerReload chan struct{}
 }
 
-// Run starts background processing to handle target updates and reload the scraping loops.
+// Run receives and saves target set updates and triggers the scraping loops reloading.
+// Reloading happens in the background so that it doesn't block receiving targets updates.
 func (m *Manager) Run(tsets <-chan map[string][]*targetgroup.Group) error {
+	go m.reloader()
 	for {
 		select {
 		case ts := <-tsets:
-			m.reload(ts)
+			m.updateTsets(ts)
+
+			select {
+			case m.triggerReload <- struct{}{}:
+			default:
+			}
+
 		case <-m.graceShut:
 			return nil
 		}
 	}
 }
 
+func (m *Manager) reloader() {
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.graceShut:
+			return
+		case <-ticker.C:
+			select {
+			case <-m.triggerReload:
+				m.reload()
+			case <-m.graceShut:
+				return
+			}
+		}
+	}
+
+}
+
+func (m *Manager) reload() {
+	m.mtxScrape.Lock()
+	var wg sync.WaitGroup
+	for setName, groups := range m.targetSets {
+		var sp *scrapePool
+		existing, ok := m.scrapePools[setName]
+		if !ok {
+			scrapeConfig, ok := m.scrapeConfigs[setName]
+			if !ok {
+				level.Error(m.logger).Log("msg", "error reloading target set", "err", fmt.Sprintf("invalid config id:%v", setName))
+				return
```

Comment:
Could it happen? Does it need to return or it could continue?

Right now the scrape manager's configuration is reloaded before the discovery manager.

https://github.com/prometheus/prometheus/blob/75bd348135019c6bf5bee4350a339964e62f2d8a/cmd/prometheus/main.go#L323-L333

So I guess it can happen when a scrape job is removed from the configuration and the configuration update has been applied to the scrape manager but the discovery manager hasn't picked up yet. I'd say that we can silently drop the item then.

_I see that this code already exists in the previous version so feel free to ignore_

Reply:
If someone changes the workflow somewhere we might hit this so this way we will know soon enough that something is not right. This is one of the problems with the current implementation of how wet handle the config reloading. 
I have suggested how to improve, but there was no feedback at the time.

Reply:
Either way this is story for another PR ;)